File: java/bypstest/src/com/wilutions/byps/test/TestRemoteStreams.java
package com.wilutions.byps.test;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.util.Map;
import java.util.TreeMap;

import junit.framework.Assert;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import com.wilutions.byps.BContentStream;
import com.wilutions.byps.BException;
import com.wilutions.byps.BWire;
import com.wilutions.byps.http.HConstants;
import com.wilutions.byps.test.api.BClient_Testser;
import com.wilutions.byps.test.api.remote.RemoteStreams;

/**
 * Tests for interface functions with stream types.
 * This test requires the webapp bypstest-srv to be started. 
 */
public class TestRemoteStreams {
	
	BClient_Testser client;
	RemoteStreams remote;
	private Log log = LogFactory.getLog(TestRemoteStreams.class);

	@Before
	public void setUp() throws BException, InterruptedException {
		client = TestUtilsHttp.createClient();
		remote = client.remoteStreams;
	}
	
	@After
	public void tearDown() throws InterruptedException {
		if (client != null) {
			client.done();
		}
	}
	
	/**
	 * Send and receive a stream.
	 * It's sent by HTTP chunked encoding.
	 * @throws InterruptedException
	 * @throws IOException
	 */
	@Test
	public void testRemoteStreamsOneStream() throws InterruptedException, IOException {
		log.info("testRemoteStreamsOneStream(");
		
		String str = "hello";
		InputStream istrm = new ByteArrayInputStream(str.getBytes());
		remote.setImage(istrm);
		
		InputStream istrmR = remote.getImage();
		ByteBuffer buf = BWire.bufferFromStream(istrmR);
		String strR = new String(buf.array(), buf.position(), buf.remaining());
		TestUtils.assertEquals(log, "stream", str, strR);
		
		remote.setImage(null);
		TestUtils.checkTempDirEmpty(client);
		
		log.info(")testRemoteStreamsOneStream");
	}
	
	/**
	 * Send and receive many streams.
	 * @throws InterruptedException
	 * @throws IOException
	 */
	@Test
	public void testRemoteStreamsManyStreams() throws InterruptedException, IOException {
		log.info("testRemoteStreamsOneStreams(");
		Map<Integer, InputStream> streams = new TreeMap<Integer, InputStream>();
		for (int i = 0; i < 10; i++) {
			String str = "hello-" + i;
			InputStream istrm = new ByteArrayInputStream(str.getBytes());
			streams.put(i, istrm);
		}
		remote.setImages(streams, -1);
		TreeMap<Integer, InputStream> istrmsR = remote.getImages();
		
		TestUtils.assertEquals(log, "streams", streams, istrmsR); // Does not compare streams.
		for (int i = 0; i < 10; i++) {
			InputStream istrmR = istrmsR.get(i);
			ByteBuffer buf = BWire.bufferFromStream(istrmR);
			String strR = new String(buf.array(), buf.position(), buf.remaining());
			TestUtils.assertEquals(log, "stream", "hello-" + i, strR);
		}
		
		log.info(")testRemoteStreamsOneStreams");
	}
	
	
	/**
	 * Wrapper class for ByteArrayInputStream.
	 * - helps to check that all streams are closed after the request
	 * - helps to check that an Exception thrown in read() is correctly handled.
	 */
	private class MyInputStream extends BContentStream {
		
		private volatile boolean isClosed;
		boolean throwEx;
		boolean throwError;
		byte[] buf;
		int idx;
		
		/**
		 * Constructor
		 * @param buf	Buffer to read from
		 * @param throwEx	true, if an IOException has to be thrown
		 * @param throwError	true, if an IllegalStateException has to be thrown
		 */
		public MyInputStream(byte[] buf, boolean throwEx, boolean throwError) {
			this.buf = buf;
			this.throwEx = throwEx;
			this.throwError = throwError;
		}
		
		@Override
		public void close() throws IOException {
			isClosed = true;
			super.close();
		}
		
		@Override
		public int read() throws IOException {
			if (throwEx) {
				throw new IOException("Test Exception");
			}
			if (throwError) {
				throw new IllegalStateException("Test Error");
			}
			if (idx < buf.length) {
				return ((int)buf[idx++]) & 0xFF;
			}
			return -1;
		}

		@Override
		public String getContentType() {
			return null;
		}

		@Override
		public long getContentLength() {
			return buf.length;
		}
	}

	/**
	 * All the input streams sent are closed by the transport layer. 
	 * Streams not requested from the server must be closed too.
	 * @throws InterruptedException 
	 * @throws BException 
	 */
	@Test
	public void testRemoteStreamsCloseStreamAfterSend() throws BException, InterruptedException {
		log.info("testRemoteStreamsCloseStreamAfterSend(");
		int count = 100;
		
		for (int i = 0; i < count; i++) {
			internalTestRemoteStreamsClosed();
			
			String logLine = "testRemoteStreamsCloseStreamAfterSend-" + i;
			client.transport.wire.getTestAdapter().printServerLog(logLine);
		}

		log.info(")testRemoteStreamsCloseStreamAfterSend");
	}

	protected void internalTestRemoteStreamsClosed() throws BException,
			InterruptedException {
		int nbOfStreams = 10;
		Map<Integer, InputStream> mystreams = new TreeMap<Integer, InputStream>();
		for (int i = 0; i < nbOfStreams; i++) {
			String str = "hello-" + i;
			MyInputStream mystrm = new MyInputStream(str.getBytes(), false, false);
			mystreams.put(i, mystrm);
		}
		remote.setImages(mystreams, 1);
		
		for (int i = 0; i < nbOfStreams; i++) {
			MyInputStream istrm = (MyInputStream)mystreams.get(i);
			TestUtils.assertEquals(log, "InputStream["+i+"].isClosed, istrm=" + istrm, true, istrm.isClosed);
		}
	}
	

	/**
	 * Handle an Exception thrown in InputStream.read correctly.
	 * The Exception must be encapsulated in an BException.
	 * It must be passed to the caller. 
	 * All streams have to be closed.
	 * The server must have received an exception too.
	 * This function tests the handling of an IOException and IllegalStateException.
	 * All messages must have been finished after a while when the client connection has been closed.
	 * @throws InterruptedException
	 * @throws IOException 
	 * @see MyInputStream
	 */
	@Test
	public void testRemoteStreamsThrowExceptionOnRead() throws InterruptedException, IOException {
		log.info("testRemoteStreamsThrowExceptionOnRead(");
		
		for (int i = 0; i < 100; i++) {
			internalTestThrowExOnRead(true, false);
			internalTestThrowExOnRead(false, true);
		}
		
		client.transport.wire.cancelAllRequests();
		
		try {
			// There should be no active messages on the server after the client side has been finished.
			long[] messageIds = null;
			int keepMessageSeconds = ((int)HConstants.KEEP_MESSAGE_AFTER_FINISHED / 1000);
			int cleanupSeconds = (int)HConstants.CLEANUP_MILLIS/1000;
			int waitUntilMessagesExpired = 2 * Math.max(cleanupSeconds, keepMessageSeconds) + 1;
			for (int retry = 0; retry <  waitUntilMessagesExpired; retry++) {
				messageIds = client.transport.wire.getTestAdapter().getAcitveMessagesOnServer(false);
				if (messageIds.length == 0) break;
				Thread.sleep(1000);
			}
			TestUtils.assertEquals(log, "active messages: ", new long[0], messageIds);
		}
		finally {
			client.done();
			client = null;
		}
		log.info(")testRemoteStreamsThrowExceptionOnRead");

	}

	private void internalTestThrowExOnRead(boolean throwEx, boolean throwError) throws InterruptedException, IOException {
		log.info("internalTestThrowExOnRead(throwEx=" + throwEx + ", throwError=" + throwError);
		
		Map<Integer, InputStream> streams = new TreeMap<Integer, InputStream>();
		for (int i = 0; i < 3; i++) {
			String str = "hello-" + i;
			InputStream istrm = new MyInputStream(str.getBytes(), i == 1 && throwEx, i == 1 && throwError);
			streams.put(i, istrm);
			log.info("strm[" + i + "]=" + str);
		}
		
		// An exception is thrown in HWireClient.
		try {
			log.info("setImages...");
			remote.setImages(streams, -1);
			Assert.fail("Exception expected");
		}
		catch (BException e) {
			log.info("setImages ex=" + e + ", OK");
			String expectedMessage = throwEx ? "Test Exception" : "Test Error";
			String expectedDetails = throwEx ? "java.io.IOException: Test Exception" : "java.lang.IllegalStateException: Test Error";
			TestUtils.assertEquals(log, "Exception Code", BException.IOERROR, e.code);
			TestUtils.assertEquals(log, "Exception Message", expectedMessage, e.msg);
			TestUtils.assertEquals(log, "Exception Details", expectedDetails, e.details);
		}
		
		// The server must have received an exception too
		try {
			log.info("throwLastException...");
			remote.throwLastException();
			Assert.fail("Exception from server expected");
		}
		catch (BException e) {
			log.info("throwLastException OK, ex=" + e);
			TestUtils.assertEquals(log, "Exception Code", BException.IOERROR, e.code);
		}
		
		// All streams must have been closed
		log.info("streams must be closed");
		int nbOfClosed = 0;
		int keepMessageSeconds = ((int)HConstants.KEEP_MESSAGE_AFTER_FINISHED / 1000) + 1;
		for (int retry = 0; retry < 10 * keepMessageSeconds; retry++) {
			for (int i = 0; i < streams.size(); i++) {
				MyInputStream istrm = (MyInputStream)streams.get(i);
				log.info("InputStream[" + i + "].isClosed=" + istrm.isClosed);
				if (istrm.isClosed) nbOfClosed++;
			}
			if (nbOfClosed == streams.size()) break;
			nbOfClosed = 0;
			Thread.sleep(100);
		}
		for (int i = 0; i < streams.size(); i++) {
			MyInputStream istrm = (MyInputStream)streams.get(i);
			TestUtils.assertEquals(log, "InputStream[" + i + "].isClosed, stream=" + istrm, true, istrm.isClosed);
		}
		
		log.info(")internalTestThrowExOnRead");
	}
	
	/**
	 * Test for cloning streams inside the server.
	 * The server can read a stream as long as the request (message with ByteBuffer) is not finished.
	 * If streams have to be stored in member variables of the server objects, 
	 * they have to be cloned. 
	 * @throws InterruptedException 
	 * @throws IOException 
	 */
	@Test
	public void testRemoteStreamsCloneStream() throws InterruptedException, IOException {
		log.info("testRemoteStreamsCloneStream(");
		
		// Use byte array as buffer
		internalTestCloneStream(1);
		internalTestCloneStream(0);
		internalTestCloneStream(HConstants.INCOMING_STREAM_BUFFER / 2);
		internalTestCloneStream(HConstants.INCOMING_STREAM_BUFFER - 1);
		internalTestCloneStream(HConstants.INCOMING_STREAM_BUFFER);
		
		// Use file as buffer
		internalTestCloneStream(HConstants.INCOMING_STREAM_BUFFER + 1);
		internalTestCloneStream(HConstants.INCOMING_STREAM_BUFFER * 2);
		
		log.info(")testRemoteStreamsCloneStream");
	}

	protected void internalTestCloneStream(long nbOfBytes) throws BException,
			InterruptedException, IOException {
		
		log.info("start upload");
		InputStream istrm = new TestUtils.MyContentStream(nbOfBytes);
		remote.setImage(istrm);
		
		for (int loop = 0; loop < 5; loop++) {
			
			// remote.getImage() clones the stream.
			log.info("start download");
			InputStream istrmR = remote.getImage();
			
			log.info("compare streams");
			istrm = new TestUtils.MyContentStream(nbOfBytes);
			TestUtils.assertEquals(log, "stream[" + loop + "]", istrm, istrmR);
		}
		
		remote.setImage(null);
		TestUtils.checkTempDirEmpty(client);

	}

	
	/**
	 * Check that streams larger than 2GB can be transferred.
	 * @throws InterruptedException 
	 * @throws IOException 
	 */
	@Test
	public void testRemoteStreamsLargeStream() throws InterruptedException, IOException {
		log.info("testRemoteStreamsLargeStream(");
		
		long nbOfBytes = 3L * 1000L * 1000L * 1000L;
		
		log.info("start upload");
		InputStream istrm = new TestUtils.MyContentStream(nbOfBytes);
		remote.setImage(istrm);
		
		log.info("start download");
		InputStream istrmR = remote.getImage();
		
		log.info("compare streams");
		istrm = new TestUtils.MyContentStream(nbOfBytes);
		TestUtils.assertEquals(log, "", istrm, istrmR);
		
		remote.setImage(null);
		TestUtils.checkTempDirEmpty(client);
		
		log.info(")testRemoteStreamsLargeStream");
	}

	
}

File: csharp/bypstest/TestRemoteStreams.cs
/* USE THIS FILE ACCORDING TO THE COPYRIGHT RULES IN LICENSE.TXT WHICH IS PART OF THE SOURCE CODE PACKAGE */
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using byps;
using byps.test.api;
using byps.test.api.remote;
using System.IO;
using System.Text;
using System.Collections.Generic;
using System.Threading;

namespace bypstest
{
    [TestClass]
    public class TestRemoteStreams
    {
        Log log = LogFactory.getLog(typeof(TestRemoteStreams));
        BClient_Testser client;
        RemoteStreams remote;

        [TestInitialize]
        public void setUp()
        {
            client = TestUtilsHttp.createClient();
            remote = client.RemoteStreams;
        }

        [TestCleanup]
        public void tearDown()
        {
            client.done();
        }

        class MyContentStreamChunked : BContentStreamWrapper
        {
            public MyContentStreamChunked(Stream innerStream)
                : base(innerStream)
            {
            }

            public override String ContentType
            {
                get
                {
                    return "application/octet-stream";
                }
            }

            public override long ContentLength
            {
                get
                {
                    return -1L;
                }
            }

        }

        [TestMethod]
        public void testRemoteStreamsOneStreamChunked()
        {
            log.info("testRemoteStreamsOneStreamChunked(");
            String str = "hello";
            Stream istrm = new MyContentStreamChunked(new MemoryStream(Encoding.ASCII.GetBytes(str)));
            remote.SetImage(istrm);
            Stream istrmR = remote.GetImage();
            ByteBuffer buf = BWire.bufferFromStream(istrmR, false);
            String strR = Encoding.ASCII.GetString(buf.array(), buf.position(), buf.remaining());
            TestUtils.assertEquals(log, "stream", str, strR);
            log.info(")testRemoteStreamsOneStreamChunked");
        }

        [TestMethod]
        public void testRemoteStreamsOneStreamNotChunked()
        {
            log.info("testRemoteStreamsOneStreamNotChunked(");
            String str = "hello";
            Stream istrm = new MemoryStream(Encoding.ASCII.GetBytes(str));
            remote.SetImage(istrm);
            Stream istrmR = remote.GetImage();
            ByteBuffer buf = BWire.bufferFromStream(istrmR, false);
            String strR = Encoding.ASCII.GetString(buf.array(), buf.position(), buf.remaining());
            TestUtils.assertEquals(log, "stream", str, strR);
            log.info(")testRemoteStreamsOneStreamNotChunked");
        }

        /// <summary>
        /// Check sending/receiving correct Content-Type and file name.
        /// </summary>
        [TestMethod]
        public void testRemoteStreamsFileStream()
        {
            log.info("testRemoteStreamsFileStream(");
            String str = "hallo";
            FileInfo file = new FileInfo(Path.GetTempFileName() + " € ß.txt");
            File.WriteAllText(file.FullName, str);
            Stream istrm = new BContentStreamWrapper(file);
            remote.SetImage(istrm);
            BContentStream istrmR = (BContentStream)remote.GetImage();

            TestUtils.assertEquals(log, "Content-Type", "text/plain", istrmR.ContentType);
            TestUtils.assertEquals(log, "Content-Length", file.Length, istrmR.ContentLength);
            TestUtils.assertEquals(log, "FileName", file.Name, istrmR.FileName);

            ByteBuffer buf = BWire.bufferFromStream(istrmR, false);
            String strR = File.ReadAllText(file.FullName);
            TestUtils.assertEquals(log, "stream", str, strR);

            TestUtils.assertEquals(log, "Content-Type", "text/plain", istrmR.ContentType);
            TestUtils.assertEquals(log, "Content-Length", file.Length, istrmR.ContentLength);
            TestUtils.assertEquals(log, "FileName", file.Name, istrmR.FileName);

            file.Delete();
            log.info(")testRemoteStreamsFileStream");
        }

        [TestMethod]
        public void testRemoteStreamsManyStreams()
        {
            log.info("testRemoteStreamsOneStreams(");
            int nbOfStreams = 10;

            for (int i = 0; i < 10; i++)
            {
                internalTestRemoteStreamsManyStreams(nbOfStreams);
            }

            log.info(")testRemoteStreamsOneStreams");
        }

        private void internalTestRemoteStreamsManyStreams(int nbOfStreams)
        {
            IDictionary<int, Stream> streams = new Dictionary<int, Stream>();
            for (int i = 0; i < nbOfStreams; i++)
            {
                String str = "hello-" + i;
                Stream istrm = new MemoryStream(Encoding.ASCII.GetBytes(str));
                streams[i] = istrm;
            }
            remote.SetImages(streams, -1);
            Dictionary<int, Stream> istrmsR = remote.GetImages();

            TestUtils.assertEquals(log, "streams", streams, istrmsR); // Does not compare streams.
            for (int i = 0; i < nbOfStreams; i++)
            {
                Stream istrmR = istrmsR[i];
                ByteBuffer buf = BWire.bufferFromStream(istrmR, false);
                String strR = Encoding.ASCII.GetString(buf.array(), buf.position(), buf.remaining());
                TestUtils.assertEquals(log, "stream", "hello-" + i, strR);
            }
        }

        private class MyInputStream : BContentStream
        {
            private Log log = LogFactory.getLog(typeof(MyInputStream));
            readonly byte[] buf;
            int idx;
            public bool isClosed;
            public bool throwEx;
            public bool throwError;

            public MyInputStream(byte[] buf, bool throwEx, bool throwError)
            {
                this.buf = buf;
                this.throwEx = throwEx;
                this.throwError = throwError;
            }

            public override void Close()
            {
                isClosed = true;
                base.Close();
            }

            public override long ContentLength
            {
                get
                {
                    return buf.Length;
                }
            }

            public override int Read(byte[] buffer, int offset, int count)
            {
                if (log.isDebugEnabled()) log.debug("Read(offs=" + offset + ", count=" + count);

                int ret = 0;
                if (throwEx)
                {
                    throw new IOException("Test Exception");
                }
                if (throwError)
                {
                    throw new InvalidOperationException("Test Error");
                }

                if (log.isDebugEnabled()) log.debug("idx=" + idx + ", buf.Length=" + buf.Length);
                if (idx < buf.Length)
                {
                    int i = 0;
                    for (; i < count && i + idx < buf.Length; i++)
                    {
                        buffer[i + offset] = buf[i + idx];
                    }
                    idx += i;
                    if (log.isDebugEnabled()) log.debug("idx=" + idx);
                    ret = i;
                }

                if (log.isDebugEnabled()) log.debug(")Read=" + ret);
                return ret;
            }

            public override int ReadByte()
            {
                if (throwEx)
                {
                    throw new IOException("Test Exception");
                }
                if (throwError)
                {
                    throw new InvalidOperationException("Test Error");
                }

                if (idx < buf.Length)
                {
                    return buf[idx++];
                }

                return 0;
            }

            public override bool CanRead
            {
                get { return true; }
            }

            public override bool CanSeek
            {
                get { return false; }
            }

            public override bool CanWrite
            {
                get { return false; }
            }

            public override void Flush()
            {
            }

            public override long Length
            {
                get { throw new NotImplementedException(); }
            }

            public override long Position
            {
                get
                {
                    throw new NotImplementedException();
                }
                set
                {
                    throw new NotImplementedException();
                }
            }

            public override long Seek(long offset, SeekOrigin origin)
            {
                throw new NotImplementedException();
            }

            public override void SetLength(long value)
            {
                throw new NotImplementedException();
            }

            public override void Write(byte[] buffer, int offset, int count)
            {
                throw new NotImplementedException();
            }
        }

        [TestMethod]
        public void testRemoteStreamsCloseStreamAfterSend()
        {
            log.info("testRemoteStreamsCloseStreamAfterSend(");

            IDictionary<int, Stream> streams = new Dictionary<int, Stream>();
            for (int i = 0; i < 10; i++)
            {
                String str = "hello-" + i;
                Stream istrm = new MyInputStream(Encoding.ASCII.GetBytes(str), false, false);
                streams[i] = istrm;
            }
            remote.SetImages(streams, 5);

            for (int i = 0; i < 10; i++)
            {
                MyInputStream istrm = (MyInputStream)streams[i];
                TestUtils.assertEquals(log, "InputStream.isClosed", true, istrm.isClosed);
            }

            log.info(")testRemoteStreamsCloseStreamAfterSend");
        }


        [TestMethod]
        public void testRemoteStreamsThrowExceptionOnRead()
        {
            log.info("testRemoteStreamsThrowExceptionOnRead(");

            internalTestThrowExOnRead(true, false);
            internalTestThrowExOnRead(false, true);

            log.info(")testRemoteStreamsThrowExceptionOnRead");

        }

        private void internalTestThrowExOnRead(bool throwEx, bool throwError)
        {
            log.info("internalTestThrowExOnRead(throwEx=" + throwEx + ", throwError=" + throwError);

            IDictionary<int, Stream> streams = new Dictionary<int, Stream>();
            for (int i = 0; i < 3; i++)
            {
                String str = "hello-" + i;
                Stream istrm = new MyInputStream(Encoding.ASCII.GetBytes(str), i == 1 && throwEx, i == 1 && throwError);
                streams[i] = istrm;
            }

            // An exception is thrown in HWireClient.
            try
            {
                log.info("setImages...");
                remote.SetImages(streams, -1);
                Assert.Fail("Exception expected");
            }
            catch (BException e)
            {
                log.info("setImages ex=" + e + ", OK");
                String expectedMessage = throwEx ? "Test Exception" : "Test Error";
                String expectedDetails = throwEx ? "System.IO.IOException: Test Exception" : "System.InvalidOperationException: Test Error";
                TestUtils.assertEquals(log, "Exception Code", BExceptionC.IOERROR, e.Code);
                TestUtils.assertEquals(log, "Exception Message", expectedMessage, e.Message);
                String readDetails = e.Details.Substring(0, expectedDetails.Length);
                TestUtils.assertEquals(log, "Exception Details", expectedDetails, readDetails);
            }

            {
                log.info("read streams");
                IDictionary<int, Stream> istrmsR = remote.GetImages();
                for (int i = 0; i < istrmsR.Count; i++)
                {
                    Stream istrmR = istrmsR[i];
                    ByteBuffer buf = BWire.bufferFromStream(istrmR, false);
                    String strR = Encoding.ASCII.GetString(buf.array(), buf.position(), buf.remaining());
                    log.info("strR=" + strR);
                }
            }

            // All streams must have been closed
            log.info("streams must be closed");
            for (int i = 0; i < streams.Count; i++)
            {
                MyInputStream istrm = (MyInputStream)streams[i];
                TestUtils.assertEquals(log, "InputStream.isClosed", true, istrm.isClosed);
            }

            log.info(")internalTestThrowExOnRead");
        }

        [TestMethod]
        public void testRemoteStreamsLargeStream()
        {
            if (!TestUtils.TEST_LARGE_STREAMS) return;
            log.info("testRemoteStreamsLargeStream(");

            long nbOfBytes = 0x100000000L;

            log.info("start upload");
            Stream istrm = new TestUtils.MyContentStream(nbOfBytes);
            remote.SetImage(istrm);

            log.info("start download");
            Stream istrmR = remote.GetImage();

            log.info("compare streams");
            istrm = new TestUtils.MyContentStream(nbOfBytes);
            TestUtils.assertEquals(log, "", istrm, istrmR);

            remote.SetImage(null);
            TestUtils.checkTempDirEmpty(client);

            log.info(")testRemoteStreamsLargeStream");
        }

        class BContentStreamWrapperFailOpen : BContentStreamWrapper
        {
            public BContentStreamWrapperFailOpen(Stream strmFromServer)
                : base((BContentStream)strmFromServer, 0)
            {
            }
            protected override Stream openStream()
            {
                throw new IOException("Stream should be passed without beeing read.");
            }

        }

        [TestMethod]
        public void testHandoverStream()
        {
            log.info("testHandoverStream(");

            List<Stream> streams = TestUtilsHttp.makeTestStreams();

            for (int i = 0; i < streams.Count; i++)
            {

                Stream istrm = streams[i];

                remote.SetImage(istrm);

                Stream strmFromServer = remote.GetImage();
                BContentStream streamFailOpen = new BContentStreamWrapperFailOpen(strmFromServer);

                remote.SetImage(streamFailOpen);

                for (int j = 0; j < 2; j++)
                {
                    List<Stream> estreams = TestUtilsHttp.makeTestStreams();
                    Stream estrm = estreams[i];
                    Stream rstrm = remote.GetImage();
                    TestUtils.assertEquals(log, "stream[" + i + "]=" + estrm, estrm, rstrm);
                }

                remote.SetImage(null);
                TestUtils.checkTempDirEmpty(client);
            }

            log.info(")testHandoverStream");
        }


          /**
           * Test with a stream class which properties are unavailable until the stream is opened.
           * @throws IOException 
           */
        [TestMethod]
        public void testStreamWithDeferedProperties()
        {
            log.info("testStreamWithDeferedProperties(");

            BContentStream bstrm = (BContentStream)remote.GetStreamDeferedProperies();
            TestUtils.assertEquals(log, "contentType", "", bstrm.ContentType);
            TestUtils.assertEquals(log, "contentLength", -1L, bstrm.ContentLength);
            TestUtils.assertEquals(log, "fileName", "", bstrm.FileName);
            TestUtils.assertEquals(log, "attachmentCode", 0, bstrm.AttachmentCode);

            bstrm.ensureProperties();
            TestUtils.assertEquals(log, "contentType", "application/mycontentype", bstrm.ContentType);
            TestUtils.assertEquals(log, "contentLength", 5L, bstrm.ContentLength);
            TestUtils.assertEquals(log, "fileName", "myfilename", bstrm.FileName);
            TestUtils.assertEquals(log, "attachmentCode", BContentStream.ATTACHMENT, bstrm.AttachmentCode);

            log.info(")testStreamWithDeferedProperties");
        }

          /**
           * Test with a stream class which properties are unavailable until the stream is opened.
           * The stream is passed back to the server and in turn read from the server. 
           * The stream properties must be unavailable until ensureProperties is called.
           * @throws IOException 
           */
        [TestMethod]
        public void testHandoverStreamWithDeferedProperties()
        {
            log.info("testHandoverStreamWithDeferedProperties(");

            BContentStream bstrm1 = (BContentStream)remote.GetStreamDeferedProperies();
            TestUtils.assertEquals(log, "contentType", "", bstrm1.ContentType);
            TestUtils.assertEquals(log, "contentLength", -1L, bstrm1.ContentLength);
            TestUtils.assertEquals(log, "fileName", "", bstrm1.FileName);
            TestUtils.assertEquals(log, "attachmentCode", 0, bstrm1.AttachmentCode);

            remote.SetStreamDoNotMaterialize(bstrm1);
            BContentStream bstrm = (BContentStream)remote.GetStreamDoNotClone();

            TestUtils.assertEquals(log, "contentType", "", bstrm.ContentType);
            TestUtils.assertEquals(log, "contentLength", -1L, bstrm.ContentLength);
            TestUtils.assertEquals(log, "fileName", "", bstrm.FileName);
            TestUtils.assertEquals(log, "attachmentCode", 0, bstrm.AttachmentCode);

            bstrm.ensureProperties();
            TestUtils.assertEquals(log, "contentType", "application/mycontentype", bstrm.ContentType);
            TestUtils.assertEquals(log, "contentLength", 5L, bstrm.ContentLength);
            TestUtils.assertEquals(log, "fileName", "myfilename", bstrm.FileName);
            TestUtils.assertEquals(log, "attachmentCode", BContentStream.ATTACHMENT, bstrm.AttachmentCode);

            log.info(")testHandoverStreamWithDeferedProperties");
        }

    }
}
